docs(adr): create ADR for import/export split (#205)
WirelessCar/nauth: adr/0005-split-import-and-export-crds-from-account-crd.md

> # ADR-5: Split import and export CRDs from Account CRD

Reviewer: ```suggestion
# ADR-5: Extract Import- and Export CRDs from Account CRD
```

Date: 2026-03-26

## Problem statement

The `Account` CRD is currently used to manage everything related to the account, which includes limits, imports and
exports. If any of these settings are broken, especially imports and exports, it can affect the entire account.

Splitting the import and export CRDs from the Account CRD would allow for better separation of concerns.

## Status

> In progress

Reviewer: **suggestion**: Since we have a decision if this PR is merged, then this status should align with this, i.e. this ADR is no longer In progress if merged as-is.

## Context

There exists several identified issues or improvements related to imports and exports, such as: validation of imports
and matching exports, import and export using activation tokens, support for grouping of related imports and exports (

> for example, jetstream related API topics), getting rid of need to apply imports and exports in correct order.

Reviewer: ```suggestion
for example, JetStream related API subjects).
```
I guess it should be `subjects` instead of `topics`, also there is never a "need to apply imports and exports in correct order". You can define either of them first, they simply just doesn't work until they are matched on both sides.

All these are concerns of imports and exports, and might benefit from the split as well as make the account resources
more robust and easier to manage.

It is still the responsibility of the `Account` CRD, to reconcile all the imports and exports that belong to the
account.

This ADR describes some options for splitting import / export into separate CRDs. It aims to scope the split into
features that exist today and only introduce new features if they are necessary or small enough. Other features are
deferred.

### Concerns

These concerns have been lifted during the writing of the ADR. They are not necessarily all solved by the solutions, but
they are important to keep in mind when evaluating the options.

- How to handle partial failures, for example, if one of the imports or exports is invalid, should the entire account
  be left out / invalidated or should the valid imports and exports still be applied?

> - Account JWT must be reconciled before imports and exports, otherwise the account might be left in a "broken" state.

Reviewer: ```suggestion
- Account CRD must be reconciled before imports and exports, otherwise the account might be left in a "broken" state.
```
Don't we mean "Account CRD" and not "JWT" here?

- Changes in imports and exports must be reflected in the account JWT, which means that the account JWT must be
  reconciled after any change in imports and exports.
- How to present reconciliation state in a good way in tools such as `kubectl`, `k9s` and `argoCD`?
- How to handle overlapping subjects for same target account, from one or more CRs?
- Can an account in one namespace reference exports/imports in another namespace?
- Will import/export activation tokens work with these options?
- Must use deterministic ordering of imports/exports, so that reconciliation does not go into loop.

### Out of Scope - Deferred to future ADRs

* Cross-cluster imports/exports
* import and export activation tokens

> * import and export drift detection and Approval workflows

Reviewer: ```suggestion
* import and export drift detection and approval workflows
```

## Options

See [work log](worklog/0005/options.md) for details about options.

### Option 1 - AccountImportExport CRD

Imports and exports are simply moved to a separate CRD, but they are still managed together. The new CRD has a reference
to the account. This option is in essence the same as the current state, but with imports and exports moved to a
separate CRD.

### Option 2 - Separate AccountImport and AccountExport CRDs

Same as Option 1 but with separate CRDs for imports and exports.
Better for separating import and export specific fields and concerns.

> For example `export` has `TokenReq (bool)` field, while `import` has `Token` field.

Reviewer: **suggestion**: This line feels implicit, I'd remove it entirely.

but it can lead to a large number of CRDs if there are many imports and exports.

### Option 4 - Contract-based imports/exports

`AccountExport` represents an explicit contract owned by the exporting account.
`AccountImport` represents an import binding owned by the importing account and references a specific `AccountExport`.

Instead of only defining raw import/export fragments, this option models the relationship between importer and exporter
as first-class resources. The `Account` CRD remains responsible for reconciling the final account JWT from all ready
import/export resources belonging to the account.

The key idea is that both sides describe their own part of the relationship:

* the exporter publishes what it is willing to expose
* the importer declares which export it wants to consume

This makes the relationship explicit, easier to validate, and easier to reason about than embedding raw import/export
fragments directly in `Account`.

## Decision

* Extract import/export to separate CRDs
* Deprecate the import and export fields in `Account` CRD.
* Use `Option 4 - Contract-based imports/exports` as a target vision

> * First release must support what is supported today by inline imports and exports

Reviewer: **suggestion**: I'd like to see this bullet as the second bullet, before deprecation- and target bullet from an importance- and/or chronological ordering.

# ADR-5: Split import and export CRDs from Account CRD

Date: 2026-03-26

## Problem statement
The `Account` CRD is currently used to manage everything related to the account, which includes limits, imports and exports. If any of these settings are broken, especially imports and exports, it can affect the entire account.

Splitting the import and export CRDs from the Account CRD would allow for better separation of concerns.

## Status

In progress

_Allowed values: In progress/Accepted/Superceeded/Rejected/Obsolete._

## Context
There exists several identified issues or improvements related to imports and exports, such as: validation of imports and matching exports, import and export using activation tokens, support for grouping of related imports and exports (for example, jetstream related API topics), getting rid of need to apply imports and exports in correct order.

> All these are concerns of imports and exports, and would benefit from the split as well as make the account resources more robust and easier to manage.

Reviewer: Now this is more considered being a nitpick rather than an actual issue. Focusing on the assumption of "easier to manage" is a minor, neglectable comment. Closing.

account.

This ADR describes some options for splitting import / export into separate CRDs. It aims to scope the split into
features that exist today and only introduce new features if they are necessary or small enough. Other features are
deferred.

### Out of Scope - Deferred to future ADRs

* Cross-cluster imports/exports
* import and export activation tokens
* import and export drift detection and Approval workflows

### Concerns

These concerns have been lifted during the writing of the ADR. They are not necessarily all solved by the solutions, but
they are important to keep in mind when evaluating the options.

- How to handle partial failures, for example, if one of the imports or exports is invalid, should the entire account
  be left out / invalidated or should the valid imports and exports still be applied?
- Account JWT must be reconciled before imports and exports, otherwise the account might be left in a "broken" state.
- Changes in imports and exports must be reflected in the account JWT, which means that the account JWT must be
  reconciled after any change in imports and exports.

> - How to present the data in a good way in `kubectl` and `k9s`?

Reviewer 2: Independent on tool used
```suggestion
- How to present the data in a good way in the custom resource?
```

Reviewer: **suggestion**: The concern was not really about making the data look nice but rather ensure binary state for each custom resource, e.g. green/red in visualization tools like kubectl, k9s or ArgoCD.

    rules:
        -   name: "orders"          # human-readable name
            subject: "orders.>"     # example: orders.<region>.<id> (orders.eu-west-1.12345) 
            type: service           # service/stream
            responseType: Singleton # service response type
        -   name: "stock-info"
            subject: "stock-info"
            type: stream
```

```yaml

AccountImport:
    name: account-a-eu-orders         # unique name
    namespace: namespace-b
    accountName: Account B            # parent account, expected to be in same namespace
    exportRef: # reference to the export this import wants to consume
        name: orders-and-stock-info   # must match the export name
        namespace: namespace-a
    exportAccountName: Account A      # expected exporting account, used for validation
    ruleBindings:
        -   name: "orders eu-west-1"        # human-readable name
            subject: "orders.eu-west-1.>"   # must match or be a subset of an export rule subject
            type: service                   # must match the export rule type
            localSubject: "eu.orders.>"     # optional local subject 
```

In this example:

> * Account A says: "I export these subjects and I allow any account to import them"

Reviewer 2: **issue:** There is nothing really bringing up the topic on how to handle private exports in this context.

Reviewer 2: I see that they are present in the worklog, but if you want an ADR then this is something which is good to have here.

Author: That was scoped out from this ADR (including using activation tokens). Initially we have the same as today, only public exports. Later we can introduce private exports, but it needs the activation token handling to be secure.

Reviewer: Private exports deferred from this ADR. Resolving.

This makes partial failures easier to reason about:

* if one import is invalid, that import becomes `Ready=False`
* any invalid rule makes the entire resource `Ready=False`
* other valid imports and exports can still remain usable
* the final account state only includes resources that are ready

For example, if an import references a subject that does not exist as whole or subset in the export, that import does
not become active.

#### How conflicts are handled

Because multiple imports and exports may contribute to the final account state, the controller must resolve them
deterministically.

This includes:

* deterministic ordering of imports and exports
* conflict detection when subjects overlap
* clear reporting when two resources cannot be merged safely

To identify conflicts, we will use nats.io's JWT imports validation function.

#### Benefits:

* Imports and Exports can be created in any order.
* Allows for exporting of group of related subjects, for example, all JetStream API subjects.
* Open to future improvements such as restricted exports (handling activation tokens) and approval workflows.

#### Drawbacks:

> * More complex controller logic to handle the relationships and validation

Reviewer 2: **note:** All of the options would require that the account is the one that reconciles, so not sure this is more complex.

Author: There is some more validations and checks that an import matches an export.

Reviewer: I agree with Henri this this option is a bit more complex regarding conciliation than the other options, due to the contractual nature of the relationship between Import and Export. Resolving.

fragments directly in `Account`.

```yaml
# simplified example
AccountExport:
  name: js-api
  namespace: namespace-a
  accountRef:
    name: Account A
    namespace: namespace-a
  rules:
    - name: info
      subject: "$JS.API.INFO"
      type: service

AccountImport:
  name: use-account-a-js-api
  namespace: namespace-b
  accountRef:
    name: Account B
    namespace: namespace-b
  exporterAccountRef:
    name: Account A
    namespace: namespace-a
  exportRef:
    name: js-api
    namespace: namespace-a
  ruleBindings:
    - name: info
```

In this example:

* Account A says: "I export this capability"
* Account B says: "I want to import that capability"
* NAuth only renders it into the final account JWT when the references and rules are valid

#### How mutual consent works

Both sides must explicitly allow the relationship.

The export must allow the importing account, and the import must explicitly reference both the export and the exporting
account. This makes consent explicit and avoids accidental or implicit trust relationships.

```yaml
AccountExport:
  access:
    mode: restricted
    allowedAccountRefs:
      - name: Account B
        namespace: namespace-b
```

> With this model, an `AccountImport` from Account B is not enough on its own. The referenced `AccountExport` must also

Reviewer: In the latest PR, activation tokens are deferred to the future, hence this comment can be ignored.

`AccountImport` binds to an `AccountExport` (`exportRef`) and asserts the expected exporting account
(`exporterAccountRef`).

`ruleBindings[].expected` declares optional invariants that the referenced export rule must satisfy.
If any expected field does not match the resolved export rule, the `AccountImport` is not ready.

`updatePolicy` does not affect validation semantics. It only controls whether the import automatically adopts upstream
export changes that continue to satisfy the declared expectations.

This means compatible upstream changes can be handled in two ways:

* `automatic`: the import adopts the updated export state automatically
* `manual`: drift is detected and surfaced, and the import stops advancing until the change is explicitly accepted

This gives a clearer operational model for upstream changes without weakening validation.

#### How conflicts are handled

Because multiple imports and exports may contribute to the final account state, the controller must resolve them
deterministically.

This includes:

* deterministic ordering of imports and exports
* conflict detection when subjects overlap
* clear reporting when two resources cannot be merged safely

> For example, if two imports resolve to the same local subject, the controller must report that conflict instead of

Reviewer: Fixed. The text now mentions relying on JWT v2 lib for validation.

#### How cross-namespace references fit

This option makes cross-namespace references technically possible because resources explicitly reference each other.

```yaml
exportRef:
  name: jetstream-api
  namespace: namespace-a
```

However, this also introduces tenancy and policy questions. The API and controller must define clearly whether
cross-namespace references are allowed, and under which rules.

So this option provides a natural place for those rules, but it still requires explicit policy decisions.

#### Why this is a good foundation for future features

Because importer/exporter relationships are explicit and referenceable, this model creates a better foundation for
future improvements such as:

* activation token handling
* approval workflows
* richer policy around import/export contracts

These features are not decided by this ADR, but Option 4 makes them easier to add later in a consistent way.

Due to the explicit referencing, import/export token validation can be implicitly handled by NAuth.

#### Full example

```yaml
AccountExport:
    name: jetstream-api
    namespace: namespace-a
    accountRef:
        name: Account A           # expected to be in same namespace

>         namespace: namespace-a

Reviewer: This has been considered in latest changes. Resolved.

    access:
        mode: restricted       # restricted = only allowed accounts can import, public = any account can import
        allowedAccountRefs:    # only needed if mode is restricted, otherwise ignored
        -   name: Account B

>             namespace: namespace-b

Reviewer: Considered in latest PR changes. Resolved due to deferred to future.

    rules:
        -   name: info              # unique name for reference in import
            subject: "$JS.API.INFO"
            type: service
            responseType: single
        -   name: stream-info
            subject: "$JS.API.STREAM.INFO.*"
            type: service
            responseType: single
        -   name: with-extra-options
            subject: "my.subject.>"
            type: stream
            options: # this is optional extra parameters (example struct below)
                advertise: true
                allowTrace: true  
```

> `rules[].name` must be unique within the export and is used for reference in the import.

Reviewer: This document is now considered pure work log notes, hence ignoring this issue for the sake of the ADR (which mentions that subject is the main identifier).

```go
package v1alpha2

type AccountExportRule struct {
    Name         string       `json:"name,omitempty"`
    Subject      Subject      `json:"subject,omitempty"`
    Type         ExportType   `json:"type,omitempty"`
    ResponseType ResponseType `json:"responseType,omitempty"`

>     ExportOptions *ExportOptions `json:"options,omitempty"`

Reviewer: Considered due to the flattening of the structures. Resolved.

}

type ExportOptions struct {
    Revocations          RevocationList   `json:"revocations,omitempty"`
    AccountTokenPosition *uint            `json:"accountTokenPosition,omitempty"`
    Advertise            *bool            `json:"advertise,omitempty"`
    ResponseThreshold    *metav1.Duration `json:"responseThreshold,omitempty"`
    AllowTrace           *bool            `json:"allowTrace,omitempty"`
    Latency              *ServiceLatency  `json:"latency,omitempty"`
}

```

And the corresponding import:

```yaml
AccountImport:
    name: account-a-jetstream
    namespace: namespace-b
    accountRef:
        name: Account B
        namespace: namespace-b
    exporterAccountRef:
        name: Account A
        namespace: namespace-a
    exportRef:
        name: jetstream-api
        namespace: namespace-a
    updatePolicy: automatic             # automatic (default) = advance status when export changes compatibly, 

>                                         # manual = detect drift and stop advancing until spec says to accept it

Reviewer: Deferred to future in current PR. Resolved.

    localSubjectPrefix: ext.account-a   # optional prefix to add to all imported subjects, for example, to avoid conflicts with local subjects
    ruleBindings:   # required, non-empty list
        -   name: info                  # export rule name
            expected:                       # validation only fields 

>                 subject: "$JS.API.INFO"     # validate remote exported subject

Reviewer: This has been flattened out in latest PR. Resolved.
